File: modules/audio-noisy-module/android/src/main/java/expo/modules/audionoisymodule/AudioNoisyModule.kt
package expo.modules.audionoisymodule

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.media.AudioManager
import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * AudioNoisyModule - Handles Android AUDIO_BECOMING_NOISY broadcast
 *
 * This broadcast is sent when audio is about to become "noisy" - typically when
 * headphones are unplugged. Best practice is to pause audio playback when this happens.
 *
 * iOS handles this natively through AVAudioSession's route change notification
 * with oldDeviceUnavailable reason.
 */
class AudioNoisyModule : Module() {
  private var noisyAudioReceiver: BroadcastReceiver? = null
  private var isListening = false

  override fun definition() = ModuleDefinition {
    Name("AudioNoisyModule")

    // Event that will be sent to JS when audio becomes noisy
    Events("onAudioBecomingNoisy")

    // Start listening for audio becoming noisy broadcasts
    AsyncFunction("startListening") {
      startListeningInternal()
      mapOf("success" to true)
    }

    // Stop listening for broadcasts (cleanup)
    AsyncFunction("stopListening") {
      stopListeningInternal()
      mapOf("success" to true)
    }

    // Check if currently listening
    Function("isListening") {
      isListening
    }

    // Lifecycle: Stop listening when module is destroyed
    OnDestroy {
      stopListeningInternal()
    }
  }

  private fun startListeningInternal() {
    if (isListening) return

    val context = appContext.reactContext ?: return

    noisyAudioReceiver = object : BroadcastReceiver() {
      override fun onReceive(context: Context?, intent: Intent?) {
        if (intent?.action == AudioManager.ACTION_AUDIO_BECOMING_NOISY) {
          // Send event to JS
          sendEvent("onAudioBecomingNoisy", mapOf(
            "reason" to "headphones_unplugged",
            "timestamp" to System.currentTimeMillis()
          ))
        }
      }
    }

    val filter = IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY)

    // Register receiver - use RECEIVER_NOT_EXPORTED for Android 13+
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      context.registerReceiver(noisyAudioReceiver, filter, Context.RECEIVER_NOT_EXPORTED)
    } else {
      context.registerReceiver(noisyAudioReceiver, filter)
    }

    isListening = true
  }

  private fun stopListeningInternal() {
    if (!isListening) return

    val context = appContext.reactContext ?: return

    try {
      noisyAudioReceiver?.let {
        context.unregisterReceiver(it)
      }
    } catch (e: Exception) {
      // Receiver might not be registered, ignore
    }

    noisyAudioReceiver = null
    isListening = false
  }
}

File: modules/audio-noisy-module/src/AudioNoisyModule.ts
import { requireNativeModule, EventEmitter, Subscription } from 'expo-modules-core';
import { Platform } from 'react-native';

// Event data interface
export interface AudioNoisyEvent {
  reason: 'headphones_unplugged';
  timestamp: number;
}

interface AudioNoisyModuleInterface {
  startListening(): Promise<{ success: boolean }>;
  stopListening(): Promise<{ success: boolean }>;
  isListening(): boolean;
}

// Only available on Android - iOS handles this natively
const isAndroid = Platform.OS === 'android';

// Load native module (Android only)
const NativeAudioNoisyModule = isAndroid
  ? requireNativeModule<AudioNoisyModuleInterface>('AudioNoisyModule')
  : null;

// Event emitter for native events
const emitter = isAndroid && NativeAudioNoisyModule
  ? new EventEmitter(NativeAudioNoisyModule)
  : null;

/**
 * Start listening for audio becoming noisy events.
 * On Android, this registers a BroadcastReceiver for ACTION_AUDIO_BECOMING_NOISY.
 * On iOS, this is a no-op (expo-audio handles route changes natively).
 *
 * @returns Promise<boolean> - true if listening started successfully
 */
export async function startListening(): Promise<boolean> {
  if (!isAndroid || !NativeAudioNoisyModule) {
    // iOS handles this natively through expo-audio
    return true;
  }

  try {
    const result = await NativeAudioNoisyModule.startListening();
    return result.success;
  } catch (error) {
    console.warn('[AudioNoisyModule] Failed to start listening:', error);
    return false;
  }
}

/**
 * Stop listening for audio becoming noisy events.
 *
 * @returns Promise<boolean> - true if stopped successfully
 */
export async function stopListening(): Promise<boolean> {
  if (!isAndroid || !NativeAudioNoisyModule) {
    return true;
  }

  try {
    const result = await NativeAudioNoisyModule.stopListening();
    return result.success;
  } catch (error) {
    console.warn('[AudioNoisyModule] Failed to stop listening:', error);
    return false;
  }
}

/**
 * Check if currently listening for audio noisy events.
 *
 * @returns boolean - true if currently listening
 */
export function isListening(): boolean {
  if (!isAndroid || !NativeAudioNoisyModule) {
    return false;
  }

  try {
    return NativeAudioNoisyModule.isListening();
  } catch {
    return false;
  }
}

/**
 * Add listener for audio becoming noisy events.
 * Called when headphones are unplugged on Android.
 *
 * @param callback - Function to call when audio becomes noisy
 * @returns Subscription - Call .remove() to unsubscribe
 *
 * @example
 * ```ts
 * const subscription = addAudioNoisyListener((event) => {
 *   console.log('Headphones unplugged:', event.timestamp);
 *   // Pause audio playback
 * });
 *
 * // Later: cleanup
 * subscription.remove();
 * ```
 */
export function addAudioNoisyListener(
  callback: (event: AudioNoisyEvent) => void
): Subscription {
  if (!emitter) {
    // Return no-op subscription for iOS
    return {
      remove: () => {},
    };
  }

  return emitter.addListener('onAudioBecomingNoisy', callback);
}

export default {
  startListening,
  stopListening,
  isListening,
  addAudioNoisyListener,
};
